### ISSUE TITLE
support custom grammared sequence types

### ISSUE THREAD

**Issue by jairideout** [biocore/scikit-bio#1175]
Changes:

- [x] rename  `skbio.sequence._iupac_sequence.IUPACSequence` class to `GrammaredSequence`: it is a sequence that conforms to the scikit-bio sequence grammar (based on IUPAC's notion of degenerates, gaps, nondegenerates, etc. but not necessarily constrained to the IUPAC alphabet).

- [x] make `GrammaredSequence` publicly importable from `skbio.sequence`

- [x] document `GrammaredSequence` as a place for users to inherit from to create grammared sequences with custom alphabets (add a doctest example of this too).

- [x] for `GrammaredSequence` make a metaclass that performs validation of its alphabet (e.g., `default_gap_char` should be present in `gap_chars`; degenerates must expand into valid nondegenerates; etc.)

- [x] make it impossible to inherit from DNA, RNA, and Protein and document why

**Comment by anderspitman** [comment]
@jairideout @ebolyen What exactly is meant by inheriting from `GrammaredSequence`? For example, say we have:

```python
class GrammaredSequence(type):
    def __new__(cls, name, parents, dct):
        if dct['default_gap_char'] not in dct['gap_chars']:
            raise Exception("...")
        return super(GrammaredSequence, cls).__new__(cls, name, parents, dct)

class CustomSequence(object, metaclass=GrammaredSequence):
    gap_chars = '-.'
    default_gap_char = '.'
```

That works fine. However, `IUPACSequence` currently also has inheritable behavior associated with it, for example the `_validate` method. So we need a way to pass on that functionality. Here's one way:

```python
class GrammaredSequence(type):
    def __new__(cls, name, parents, dct):
        if dct['default_gap_char'] not in dct['gap_chars']:
            raise Exception("...")

        dct['_validate'] = cls._validate

        return super(GrammaredSequence, cls).__new__(cls, name, parents, dct)
    
    def _validate(cls):
        pass
```

That feels really hacky and I wouldn't be surprised if there are serious consequences I'm not aware of. Another solution would be to have a separate class for behavior:

```python
class GrammaredSequenceMeta(type):
    def __new__(cls, name, parents, dct):
        if dct['default_gap_char'] not in dct['gap_chars']:
            raise Exception("...")
        return super(GrammaredSequenceMeta, cls).__new__(cls, name, parents, dct)
    
class GrammaredSequence(object):
    def _validate(cls):
        pass

class CustomSequence(GrammaredSequence, metaclass=GrammaredSequenceMeta):
    gap_chars = '-.'
    default_gap_char = '.'
```

I like this better but still not ideal. A third approach is sort of a combination of the first 2:

```python
class GrammaredSequenceMeta(type):
    def __new__(cls, name, parents, dct):
        if dct['default_gap_char'] not in dct['gap_chars']:
            raise Exception("...")
            
        parents = tuple([GrammaredSequence])
        
        return super(GrammaredSequenceMeta, cls).__new__(cls, name, parents, dct)
    
class GrammaredSequence(object):
    def _validate(cls):
        pass

class CustomSequence(metaclass=GrammaredSequenceMeta):
    gap_chars = '-.'
    default_gap_char = '.'
```

With that the metaclass injects the parent which contains the behavior. I like this the most, but I still feel like there might be a better way. Ideally someone could simply inherit `GrammaredSequence` without ever knowing it was a metaclass. Do you guys know of anything?

**Comment by jairideout** [comment]
Can you make a private metaclass (e.g. `GrammaredSequenceMeta`) that performs alphabet/grammar validation and make `GrammaredSequence` use that metaclass? Then users won't have to know/care about the metaclass (because it is private) and they can just inherit from `GrammaredSequence` as usual.

**Comment by jairideout** [comment]
Sorry for the confusion, the issue description is unclear about `GrammaredSequence` *having* a metaclass that performs validation. I'll fix that.

**Comment by anderspitman** [comment]
It looks like that will work. The only caveat I see is that `GrammaredSequence` must have valid defaults defined for things to be validated, but that doesn't seem like too much of a problem.

**Comment by ebolyen** [comment]
Would it be possible to inherit from `ABCMeta` for `GrammaredSequenceMeta` and only do the checks after `super` is done? Then the ABC can complain, and when it's done we get to as well.

**Comment by anderspitman** [comment]
I was just looking at how to resolve the multiple metaclass problem. We can try chaining [`six.add_metaclass`](http://pythonhosted.org/six/#six.add_metaclass) or maybe something like [this](http://stackoverflow.com/a/14718529/943814).

**Comment by anderspitman** [comment]
Seems to work:

```python
class Meta(type):
    def __new__(cls, name, parents, dct):
        print("meta called")
        
        return super(Meta, cls).__new__(cls, name, parents, dct)

class OtherMeta(type):
    def __new__(cls, name, parents, dct):
        print("other called")
        
        return super(OtherMeta, cls).__new__(cls, name, parents, dct)
    
@six.add_metaclass(OtherMeta)
@six.add_metaclass(Meta)
class MyType(object):
    pass

meta called
other called
```

**Comment by anderspitman** [comment]
Works with `ABCMeta`, too. Makes me wonder, why does `ABCMeta` raise exceptions at object instantiation time rather than class declaration time?

**Comment by ebolyen** [comment]
Cool! `ABCMeta` probably behaves that way because you could decorate properties onto the class-object prior to any instantiation, so the ABC might be fulfilled after class declaration (or perhaps even in response to failure).

**Comment by anderspitman** [comment]
Does it still make sense for lowercase characters to be invalid by default for any `GrammaredSequence`? Seems like not.

**Comment by ebolyen** [comment]
hmm, I think you are right, but some of our writers will call `lowercase` so a `GrammaredSequence` would need to have that method at least, maybe it just wouldn't do anything?

**Comment by anderspitman** [comment]
Fine with me. If this becomes an issue at some point we can deal with it then.

**Comment by ebolyen** [comment]
@anderspitman @jairideout pointed out that this shouldn't be an issue at all in practice, if a lowercase letter is in the alphabet then it will pass `_validate`. So just don't have the metaclass care about casing.

**Comment by anderspitman** [comment]
@jairideout @ebolyen The metaclasses don't seem to be playing nice with `classproperty`. When I try to access any of the properties from the metaclass within `__new__` or  `__init__` I'm getting: 

`TypeError: argument of type 'classproperty' is not iterable`

**Comment by anderspitman** [comment]
Can we develop a more complete list of validations to perform? Other than default gap and degenerate expansion?

**Comment by ebolyen** [comment]
gaps, degenerates, and nondegenerates do not appear in each other's sets

I think that is probably it.

**Comment by anderspitman** [comment]
Will do.

**Comment by anderspitman** [comment]
Also, I think we need to make classproperty publicly importable in order for people to inherit from `GrammaredSequence`

**Comment by anderspitman** [comment]
Either that or automagically decorate the appropriate attributes from the subclass

**Comment by ebolyen** [comment]
I'm ok with that, but static properties work fine too.
```python
class MySequence(GrammaredSequence):
    degenerate_map = {'X': set(['A', 'B', 'C')}
    nondegenerate_chars = set(['A', 'B', 'C'])
```

If the user is creating their own class, we can trust them to understand what mutation will do.
We use `classproperty` to avoid situations (which have happened) where the alphabet will be borrowed and then mutated for a different purpose without realising the original on the class has been modified.

**Comment by anderspitman** [comment]
The problem is that the error if they forget to make it static or `classproperty` is very vague: "argument of type 'function' is not iterable".

I'll make `classproperty` public and document to use that or static properties.

**Comment by ebolyen** [comment]
:+1:

**Comment by anderspitman** [comment]
Here's my initial approach for disabling subclassing:

```python
class DisableSubclassingMeta(type):
    def __new__(mcls, name, parents, dct):
        for parent in parents:
            if getattr(parent, '__subclassing_disabled__', False):
                raise Exception("Subclassing is disabled for class %s" % parent.__name__)
        cls = super(DisableSubclassingMeta, mcls).__new__(mcls, name, parents, dct)
        cls.__subclassing_disabled__ = True
        return cls
```

This works great on its own, but unfortunately applying that metaclass is a mess due to trying to combine `ABCMeta`, `GrammaredSequenceMeta`, and `DisableSubclassingMeta` all at once. It's running into MRO issues. Any suggestions?

**Comment by ebolyen** [comment]
Here is another example which doesn't use a dunder property:
http://stackoverflow.com/a/2826746/579416

As for `mro` would it be possible to just say `with_metaclass(Final, GrammaredSequence)`? Ideally each would only need this one metaclass and the rest is inheritance. I guess I don't know how you are applying this.

**Comment by anderspitman** [comment]
Using the current tip of #1219 the following doesn't work:

```python
class MyClass(with_metaclass(Final, GrammaredSequence)):
    pass
```

First because `Final` isn't inheriting `ABCMeta`, `GrammaredSequenceMeta`, and `type`. Changing the declaration of `Final` to:

```python

class Final(ABCMeta, GrammaredSequenceMeta, type):
    ...
```

Changes the error to MRO resolution. But I haven't been able to figure out a solution. Probably will end up needing to structure our metaclass inheritance very carefully. I'll keep working on it...

**Comment by anderspitman** [comment]
See the section on "Conflicting Metaclasses" [here](http://gnosis.cx/publish/programming/metaclass_2.html).

**Comment by ebolyen** [comment]
What if you used `add_metaclass`? It should reconstruct the classobj after GrammaredSequence has done it's thing?

If the `MRO` is still an issue, skip using super and call `__new__` explicitly on the `typeclass`

**Comment by ebolyen** [comment]
Thanks for the link! I'll check it out in a few.

**Comment by ebolyen** [comment]
@anderspitman We're going to bump this to the next release, so there's a ton of time to figure this out now.

**Comment by anderspitman** [comment]
Sounds good.

**Comment by anderspitman** [comment]
@jairideout @ebolyen finally looked at this again. The MRO problem was easily fixed by inheriting only from `GrammaredSequenceMeta`. Currently all requested functionality has been implemented. I just need to add tests and docs for the last bit. Oh yeah and merge...

One last thing. I feel like I already asked this at one point but can't remember: why exactly are we not allowing subclassing of `DNA`, `RNA`, and `Protein`?

**Comment by jairideout** [comment]
Awesome, thanks @anderspitman! We're not allowing subclassing of `DNA`, `RNA`, and `Protein` because some of their methods are hardcoded to use a specific alphabet (IUPAC), so subclasses that redefine the alphabet may not work correctly.

**Comment by jairideout** [comment]
Fixed in #1273.
